pkg/node/snapshot.go

```diff
@@ -0,0 +1,153 @@
+}
+
+func (f *SnapshotLogFilterer) loadAndProcessSnapshot() error {
+	f.logger.Info("loading batch snapshot during construction")
+	data := archive.GetBatchSnapshot(true)
+
+	dataReader := bytes.NewReader(data)
+	gzipReader, err := gzip.NewReader(dataReader)
+	if err != nil {
+		f.logger.Error(err, "failed to create gzip reader for batch import")
+		return fmt.Errorf("create gzip reader: %w", err)
+	}
+	defer gzipReader.Close()
+
+	if err := f.parseLogs(gzipReader); err != nil {
+		f.logger.Error(err, "failed to parse logs from snapshot")
+		return err
+	}
+
+	f.isLoaded = true
+	f.logger.Info("batch snapshot loaded successfully during construction", "log_count", len(f.loadedLogs), "max_block_height", f.maxBlockHeight)
+	return nil
+}
+
+func (f *SnapshotLogFilterer) parseLogs(reader io.Reader) error {
+	var parsedLogs []types.Log
+	var currentMaxBlockHeight uint64
+	scanner := bufio.NewScanner(reader)
+
+	for scanner.Scan() {
+		line := scanner.Bytes()
+		if len(bytes.TrimSpace(line)) == 0 {
+			continue
+		}
+
+		var logEntry types.Log
+		if err := json.Unmarshal(line, &logEntry); err != nil {
+			f.logger.Warning("failed to unmarshal log event, skipping line", "error", err, "line_snippet", string(line[:min(len(line), 100)]))
+			continue
+		}
+
+		if logEntry.BlockNumber > currentMaxBlockHeight {
+			currentMaxBlockHeight = logEntry.BlockNumber
+		}
+		parsedLogs = append(parsedLogs, logEntry)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return fmt.Errorf("error scanning batch import data: %w", err)
+	}
+
+	f.loadedLogs = parsedLogs
+	f.maxBlockHeight = currentMaxBlockHeight
+	f.isLoaded = true
+	return nil
+}
+
+func (f *SnapshotLogFilterer) FilterLogs(ctx context.Context, query ethereum.FilterQuery) ([]types.Log, error) {
+	f.logger.Debug("filtering pre-loaded logs", "total_logs", len(f.loadedLogs), "query", query)
+
+	filtered := make([]types.Log, 0, len(f.loadedLogs))
+
+	for _, log := range f.loadedLogs {
+		if query.FromBlock != nil && log.BlockNumber < query.FromBlock.Uint64() {
+			continue
+		}
+		if query.ToBlock != nil && log.BlockNumber > query.ToBlock.Uint64() {
+			continue
+		}
+
+		if len(query.Addresses) > 0 {
+			addressMatch := slices.Contains(query.Addresses, log.Address)
+			if !addressMatch {
+				continue
+			}
+		}
+
+		if len(query.Topics) > 0 {
+			topicMatch := true
+
+			for i := 0; i < len(query.Topics) && i < 4; i++ {
+				if i >= len(query.Topics) || len(query.Topics[i]) == 0 {
+					continue
+				}
+
+				if i >= len(log.Topics) {
+					topicMatch = false
+					break
+				}
+
+				hasMatch := slices.Contains(query.Topics[i], log.Topics[i])
+
+				if !hasMatch {
+					topicMatch = false
+					break
+				}
+			}
+
+			if !topicMatch {
+				continue
+			}
+		}
+
+		filtered = append(filtered, log)
+	}
+
+	f.logger.Debug("filtered logs", "input_count", len(f.loadedLogs), "output_count", len(filtered))
+	return filtered, nil
+}
+
+func (f *SnapshotLogFilterer) BlockNumber(_ context.Context) (uint64, error) {
```

Comment:
underscore can be removed `BlockNumber(context.Context) (uint64, error)`

Reply:
+1